### PR TITLE
Bug Fix Issue #4603 Access problem

### PR DESCRIPTION
**Pull request by OSobky**
## What changes are proposed in this pull request?

(Please fill in changes proposed in this fix)
as discussed in #4603 there is permission error whenever docker or kubernetes are used.

I added the `onerror` function. Which `shutil.rmtree` will use whenever there is permission error to give it access to removing the temp directory. 

## How is this patch tested?

Successfully used MLflow projects with Docker and Kubernetes

(Details)

### Is this a user-facing change?

- [x] No. You can skip the rest of this section.
- [ ] Yes. Give a description of this change to be included in the release notes for MLflow users.

(Details in 1-2 sentences. You can just refer to another PR with a description if this PR is part of a larger change.)

### What component(s), interfaces, languages, and integrations does this PR affect?
Components 
- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [x] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: Local serving, model deployment tools, spark UDFs
- [ ] `area/server-infra`: MLflow server, JavaScript dev server
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface 
- [ ] `area/uiux`: Front-end, user experience, JavaScript, plotting
- [x] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language 
- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations
- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<a name="release-note-category"></a>
### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [ ] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

### CODE REVIEW

**Comment by OSobky** [comment]
Is there any updates on this PR? is it blocked by anything?

**Comment by harupy** [comment]
@OSobky Thanks for fixing this issue! I've pushed a commit to use a more conservative implementation.

File: mlflow/projects/docker.py [hunk 1]
```diff
@@ -114,7 +139,7 @@ def _create_docker_build_ctx(work_dir, dockerfile_contents):
             output_filename=result_path, source_dir=dst_path, archive_name=_PROJECT_TAR_ARCHIVE_NAME
         )
     finally:
-        shutil.rmtree(directory)
+        shutil.rmtree(directory, onerror=handle_readonly)
```
Comment (WeichenXu123) [inline mlflow/projects/docker.py:142]: could you add a unit test to verify the handle_readonly works ?

Comment (harupy) [reply mlflow/projects/docker.py]: Done!

**Review by WeichenXu123 — APPROVED** [review]
LGTM, thanks!

**Comment by OSobky** [comment]
@harupy Sorry for the super late reply! however, the commit looks good 
Thank you for checking it and merging it 👍

**Pull request merged**